File: 355.设计推特.js
/*
 * @lc app=leetcode.cn id=355 lang=javascript
 *
 * [355] 设计推特
 */

// @lc code=start

var Twitter = function() {
    this.article = [];
    this.user = new Map();
    this.tfUser = function (userId) {
        if (!this.user.has(userId)) {
            this.user.set(userId, []);
        }
    }
};

/** 
 * @param {number} userId 
 * @param {number} tweetId
 * @return {void}
 */
Twitter.prototype.postTweet = function(userId, tweetId) {
    this.article.unshift([userId, tweetId]);
    this.tfUser(userId)
};

/** 
 * @param {number} userId
 * @return {number[]}
 */
Twitter.prototype.getNewsFeed = function(userId) {
    this.tfUser(userId);
    let arr = this.user.get(userId);
    arr.push(userId);
    let res = [];
    this.article.forEach((values) => {
        if (arr.includes(values[0])) {
            res.push(values[1])
        }
    })
    if (res.length > 10) {
        res =  res.slice(0, 10)
    }
    return res
};

/** 
 * @param {number} followerId 
 * @param {number} followeeId
 * @return {void}
 */
Twitter.prototype.follow = function(followerId, followeeId) {
    this.tfUser(followerId);
    this.user.get(followerId).push(followeeId);
};

/** 
 * @param {number} followerId 
 * @param {number} followeeId
 * @return {void}
 */
Twitter.prototype.unfollow = function(followerId, followeeId) {
    this.tfUser(followerId);
    let index = this.user.get(followerId).indexOf(followeeId);
    if (index !== -1) {
        this.user.get(followerId).splice(index, 1);
    }
};

/**
 * Your Twitter object will be instantiated and called as such:
 * var obj = new Twitter()
 * obj.postTweet(userId,tweetId)
 * var param_2 = obj.getNewsFeed(userId)
 * obj.follow(followerId,followeeId)
 * obj.unfollow(followerId,followeeId)
 */
// @lc code=end

File: 355.设计推特.rs
/*
 * @lc app=leetcode.cn id=355 lang=rust
 *
 * [355] 设计推特
 */

// @lc code=start
use std::collections::{HashMap,HashSet};

struct Twitter {
    tweets: Vec<(i32, i32)>,
    followees: HashMap<i32, HashSet<i32>>
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Twitter {

    fn new() -> Self {
        Self {
            tweets: Vec::with_capacity(8),
            followees: HashMap::new()
        }
    }
    
    fn post_tweet(&mut self, user_id: i32, tweet_id: i32) {
        self.tweets.push((user_id, tweet_id));
    }
    
    fn get_news_feed(&self, user_id: i32) -> Vec<i32> {
        self.tweets
            .iter()
            .rev()
            .filter(|(x, _)| {
                *x == user_id
                    || self
                        .followees
                        .get(&user_id)
                        .map_or(false, |s| s.contains(x))
            })
            .map(|(_, x)| *x)
            .take(10)
            .collect()
    }
    
    fn follow(&mut self, follower_id: i32, followee_id: i32) {
        match self.followees.get_mut(&follower_id) {
            Some(s) => {
                s.insert(followee_id);
            }
            None => {
                let mut set = HashSet::new();
                set.insert(followee_id);
                self.followees.insert(follower_id, set);
            }
        }
    }
    
    fn unfollow(&mut self, follower_id: i32, followee_id: i32) {
        match self.followees.get_mut(&follower_id) {
            Some(s) => {
                s.remove(&followee_id);
            }
            None => {
                self.followees.insert(follower_id, HashSet::new());
            }
        }
    }
}

/**
 * Your Twitter object will be instantiated and called as such:
 * let obj = Twitter::new();
 * obj.post_tweet(userId, tweetId);
 * let ret_2: Vec<i32> = obj.get_news_feed(userId);
 * obj.follow(followerId, followeeId);
 * obj.unfollow(followerId, followeeId);
 */
// @lc code=end

